### DISCUSSION TITLE
Como gerenciar integridade da url para QRCode?

### DISCUSSION THREAD

**Discussion by deppbrazil** [frontendbr/forum#1130] [Perguntas e Respostas]
# Help me ❌  #

> Produzimos ebooks em html aqui na Kroton que também tem uma versão impressa, em determinada parte do conteúdo precisamos colocar um material de apoio **(exemplos de algorítimos em pdf)** e a solução foi hospedar esse material de apoio em pdf em um bucket que temos no AWS e na página do livro fazer um QRCode pelo Indesign que aponte a url do bucket do AWS até o caminho do pdf para apresentar no browser esses algorítimos. 

Então a url do pdf ficou assim: https://cm-kls-content.s3.amazonaws.com/ebook/embed/qr-code/2018-2/programacao-orientada-a-objetos-II/u1/s1/codigo.pdf

esse é o qrcode que aponta para url acima: 
[qrcode.pdf](https://github.com/frontendbr/forum/files/2149501/qrcode.pdf)

O teste rolou legal... a dúvida é se na pior das hipóteses o pessoal de infra resolve mudar o bucket e que não exista mais esse começo com https://cm-kls.content.s3.amazonaws.com então o qrcode já feito e impresso perderia sua utilidade para o conteúdo já impresso.

Como eu gerencio essa url caso ela mude? ou seja como faço um caminho genérico que essa url esteja dentro e o código qr seja desenhado através desse caminho genérico e imutável e caso **https://cm-kls-content** seja alterada não impacta no código qr já produzido?

### Ou alguma outra ideia será muito bem vinda... ###

**Comment by vicainelli** [comment]
Você pode usar algum encurtador de url e gerar o QR code com a url curta, assim consegue administrar a url original dentro dessa url curta.

**Comment by wmartins** [comment]
Eu não manjo muito de QR Code, mas acredito que isso não seja possível com o QR Code em si, já que uma mudança no valor dele mudaria também a imagem gerada. Mas aí seguem alguns questionamentos/reflexões que talvez ajudem:

1. É uma opção, caso o bucket mude, gerar novamente o QR Code?

2. É possível vocês terem um "acordo" para não mudar esse bucket nunca? 😄

Caso as opções anteriores não sejam possívels, uma forma de resolver é gerenciar isso com um HTTP Server e algum banco de dados.
Por exemplo, fazer com que o QR Code não aponte diretamente para o S3, mas sim para `https://endereco-do-http-server/{algum-id-gerado}`. Esse `algum-id-gerado` é algo que vai estar no teu banco, que deverá ter uma estrutura tipo a seguinte:

| id | url |
|---|----|
|123|https://cm-kls-content.s3.amazonaws.com/ebook/embed/qr-code/2018-2/programacao-orientada-a-objetos-II/u1/s1/codigo.pdf|

No caso do bucket mudar, tu só faz um update no banco e tudo continua funcionando.

O *downside* é que isso vai requerir um pouco mais de desenvolvimento, monitoramento, escala (porque ao invés de ir na Amazon diretamente, **todos** os teus clientes irão nesse HTTP server). Então é bom ver se os *tradeoffs* valem a pena para vocês.

**Comment by kivervinicius** [comment]
@deppbrazil caso tenha algum servidor com nginx faz um proxy-reverso, funciona mto bem, caso tenha pinga aqui que explico como fazer.

**Comment by deppbrazil** [comment]
@vicainelli pensei nessa possibilidade, me corrija se eu estiver errado... rsrs mas neste caso o encurtador é um token para a url original, caso quebre minha url original o link curto já não serve mais de nada, certo? então como gerencio isso? neste caso tem algum serviço pago para que a url curta nunca mude ou seja caso precise trocar a url original o link curto seja sempre o mesmo? 

Com isso gero o QRCode a partir dessa url curta

**Comment by deppbrazil** [comment]
@wmartins bom vamos lá:

## 1. É uma opção, caso o bucket mude, gerar novamente o QR Code? ##
Sim para os livros em html, pois é hospedado em ambiente web, o problema são para os impressos que no caso já teriam sido impressos...rsrs

## 2. É possível vocês terem um "acordo" para não mudar esse bucket nunca? ##
Então, na verdade o ideal seria nunca mudar o bucket no S3 mesmo, mas isso ta atrelado as tomadas de decisões de VP de T.I, caso futuramente eles tenham que trocar toda estrutura do aws não seria essa minha solicitação que faria o contrario, então me antecipando caso isso ocorra eu não fico dependente dessa tomada de decisão. =s
___
Gostei da opção do http server, não tinha pensado nisso, mas dessa forma tenho que ter outro servidor e um domínio configurado nele, certo? Pois no aws temos apenas o serviço de Storage do S3, é meio burocrático aqui conseguirmos um banco via rds no aws :(

**Comment by deppbrazil** [comment]
@kivervinicius antigamente até tínhamos um no kinghost para testes usando apache e mysql, mas atualmente migramos tudo para aws que é o nosso servidor de produção, e no aws somos restritos a alguns serviços ex: S3, Polly... quando se trata de banco, configuração do server fica com TI Infra. =s

**Comment by vicainelli** [comment]
@deppbrazil o bit.ly faz isso, você consegue customizar as urls mesmo no plano free.

**Comment by wmartins** [comment]
@vicainelli, não sabia que rolava fazer isso pelo bit.ly.

Dei uma olhada aqui, acho que pro @deppbrazil seria até melhor do que codificar isso (conforme a minha ideia anterior).
Fora que o bit.ly tem alguns dados de analytics, integrações, deep links e outras features. Baita dica!

Ah, e a vantagem também é que eles garantem a questão da disponibilidade. Menos uma preocupação. :)

**Comment by deppbrazil** [comment]
@vicainelli @wmartins é esse https://bitly.com/ ?

**Comment by vicainelli** [comment]
Esse mesmo @deppbrazil

**Comment by deppbrazil** [comment]
@vicainelli é possível editar a url destacada? tentei de todas as formas e não achei essa opção, apenas consigo editar o link curto customizado! =s
![screenshot_1](https://user-images.githubusercontent.com/19332407/42591254-8b438db8-851c-11e8-942e-c14db5028432.png)

**Comment by vicainelli** [comment]
Não, você não consegue editar o link do seu livro.
O que você pode fazer, caso mude o link do bucket, é modificar a url do atual, e criar um novo com o nome antigo. 

### Hoje
Você cria uma url, como está fazendo agora.
`https://cm-kls-content.s3.amazonaws.com/ebook/embed/qr-code/2018-2/programacao-orientada-a-objetos-II/u1/s1/codigo.pdf` 
`bit.ly/livro-pdf`

### Amanhã
Você altera o nome do link atual
`https://cm-kls-content.s3.amazonaws.com/ebook/embed/qr-code/2018-2/programacao-orientada-a-objetos-II/u1/s1/codigo.pdf` 
`bit.ly/livro-pdf-old`

E cria um novo com o nome antigo
`https://cm-kls-content.s3.amazonaws.com/ebook/embed/qr-code/2018-2/programacao-orientada-a-objetos-II/u1/s1/codigo-2.pdf`
`bit.ly/livro-pdf`

**Comment by deppbrazil** [comment]
@vicainelli pensei exatamente no seu exemplo acima, porém quando fiz um teste funcional me apresentou erros...

![screenshot_1](https://user-images.githubusercontent.com/19332407/42647133-75eb175c-85d9-11e8-925c-0cb1a382ea65.png)

neste exemplo acima eu fiz um link personalizado **http://bit.ly/qrcode-kls-2018-2-desenvolvimento-para-dispositivos-moveis-u1s3** e apontei a url **https://cm-kls-content.s3.amazonaws.com/ebook/embed/qr-code/2018-2/desenvolvimento-para-dispositivos-moveis/u1/s3/codigo.pdf**

Depois quando precisei que o mesmo endereço curto apontasse para outra url, na atual coloquei **http://bit.ly/qrcode-kls-2018-2-desenvolvimento-para-dispositivos-moveis-u1s3-old** e criei um novo endereço curto com o nome **http://bit.ly/qrcode-kls-2018-2-desenvolvimento-para-dispositivos-moveis-u1s3** e me deu o erro acima como se esse endereço curto estivesse em uso. 

## Será que pela conta pro daria certo? pois no erro ele fala sobre essa possibilidade com upgrade. ##

**Comment by vicainelli** [comment]
O seu ` http://bit.ly/qrcode-kls-2018-2-desenvolvimento-para-dispositivos-moveis-u1s3 ` ainda está apontando pro pdf, da uma conferida pra ver se alrterou mesmo pro final `-old`

**Comment by deppbrazil** [comment]
@vicainelli sim alterei, porém não sai o vinculo do anterior sem o **-old** não sei pq

E agora as duas urls curtas apontam para o PDF =(
**http://bit.ly/qrcode-kls-2018-2-desenvolvimento-para-dispositivos-moveis-u1s3
http://bit.ly/qrcode-kls-2018-2-desenvolvimento-para-dispositivos-moveis-u1s3-old**

Acho que é por isso que não consigo criar uma nova e colocar o nome `http://bit.ly/qrcode-kls-2018-2-desenvolvimento-para-dispositivos-moveis-u1s3`, será que é um bug?

Notei também que não tem como deletar quando cria uma url curta, apenas dar um hide.... e isso gera uma sujeira no painel =x

**Comment by deppbrazil** [comment]
Acho que a funcionalidade do redirect funciona apenas em conta paga:
**https://support.bitly.com/hc/en-us/articles/230900588-How-do-I-redirect-a-bitlink-**
**https://support.bitly.com/hc/en-us/articles/230561307-My-customized-bitlink-has-already-been-taken-**

**Comment by vicainelli** [comment]
É, não tem como deletar, apenas esconder.
Mas talvez tenha sido um bug mesmo na hora de você gerar, pois da pra fazer sim, eu fiz uns testes antes de sugerir isso.

Se a conta paga for uma solução viável pra você, não fica queimando neurônio não, manda bala.

**Comment by deppbrazil** [comment]
@vicainelli fiz contato com o suporte porém os valores são inviáveis, e a versão free mto bugada, desta forma vou ter que seguir mesmo com minha url da amazon que é mais seguro, mas valew pela dica. 

## :) ##

**Comment by Recargacelular** [comment]
Olá pessoal, aproveitando o fato de que vocês estão tocando no assunto de códigos Qr, li recentemente que Pix vai funcionar com esse sistema, gostaria de saber se vocês têm informações claras sobre como vai funcionar e se acham confiável aplicá-lo a um negócio. Achei esse artigo com boas informações sobre esse assunto, mais, gostaria que alguém me explicasse melhor. Obrigado pela sua ajuda de antemão.

[Pagamento Pix com Qr Code](https://www.qr-code-pagamento.com.br/como-fazer-pagamento-pix-qr-code/)

**Comment by vicainelli** [comment]
@Recargacelular o que o os bancos do Brasil estão fazendo com o Pix é algo bem similar ao que já existe aqui na Holanda, e outros países.
Aqui o sistema é o [iDeal](https://ideal.nl) basicamente você faz uma cobrança / pagamento / transferência via QR code, eu uso bastante com o [banco que tenho conta](https://ideal.bunq.com/) aqui.
Fraudes vão ter, mas acredito que serão bem menores do que a quantidade que temos hoje com os boletos por exemplo.

**Comment by marale1994** [comment]
@Recargacelular Olá! Na verdade, o PIX trabalha com um código QR, tanto dinâmico como estático... Convido-vos a ler [este artigo](https://www.hwsinet.com/o-que-e-pix-pagamento-instantaneo-gratis/) onde há boas informações sobre este meio de pagamento!
